src/storage/authorization/oauth2.cpp

```diff
@@ -0,0 +1,168 @@
+	try {
+		string api_result = APIUtils::PostRequest(context, uri, post_data);
+		doc = std::unique_ptr<yyjson_doc, YyjsonDocDeleter>(ICUtils::api_result_to_doc(api_result));
+	} catch (std::exception &ex) {
+		ErrorData error(ex);
+		throw InvalidConfigurationException("Could not get token from %s, captured error message: %s", uri,
+		                                    error.RawMessage());
+	}
+	//! FIXME: the oauth/tokens endpoint returns, on success;
+	// { 'access_token', 'token_type', 'expires_in', <issued_token_type>, 'refresh_token', 'scope'}
+	auto *root = yyjson_doc_get_root(doc.get());
+	auto access_token_val = yyjson_obj_get(root, "access_token");
+	auto token_type_val = yyjson_obj_get(root, "token_type");
+	if (!access_token_val) {
+		throw InvalidConfigurationException("OAuthTokenResponse is missing required property 'access_token'");
+	}
+	if (!token_type_val) {
+		throw InvalidConfigurationException("OAuthTokenResponse is missing required property 'token_type'");
+	}
+	string token_type = yyjson_get_str(token_type_val);
+	if (!StringUtil::CIEquals(token_type, "bearer")) {
+		throw NotImplementedException(
+		    "token_type return value '%s' is not supported, only supports 'bearer' currently.", token_type);
+	}
+	string access_token = yyjson_get_str(access_token_val);
+	return access_token;
+}
+
+string OAuth2Authorization::GetRequest(ClientContext &context, const IRCEndpointBuilder &endpoint_builder,
+                                       curl_slist *extra_headers) {
+	return APIUtils::GetRequest(context, endpoint_builder, token, extra_headers);
+}
+
+unique_ptr<OAuth2Authorization> OAuth2Authorization::FromAttachOptions(ClientContext &context,
+                                                                       IcebergAttachOptions &input) {
+	auto result = make_uniq<OAuth2Authorization>();
+
+	unordered_map<string, Value> remaining_options;
+	string secret;
+	for (auto &entry : input.options) {
+		auto lower_name = StringUtil::Lower(entry.first);
+		if (lower_name == "secret") {
+			secret = entry.second.ToString();
+		} else if (lower_name == "oauth2_scope") {
+			result->scope = entry.second.ToString();
+		} else if (lower_name == "oauth2_server_uri") {
+			result->uri = entry.second.ToString();
+		} else if (lower_name == "client_id") {
+			result->client_id = entry.second.ToString();
+		} else if (lower_name == "client_secret") {
+			result->client_secret = entry.second.ToString();
+		} else if (lower_name == "oauth2_grant_type") {
+			result->grant_type = entry.second.ToString();
+		} else {
+			remaining_options.emplace(std::move(entry));
+		}
+	}
+
+	Value token;
+	// Check if any of the options are given that indicate intent to provide options inline, rather than use a secret
+	bool user_intends_to_use_secret = true;
+	if (!result->scope.empty() || !result->uri.empty() || !result->client_id.empty() ||
+	    !result->client_secret.empty()) {
+		user_intends_to_use_secret = false;
```

Comment:
I think this can be much less complicated
We can just put apart all the options for creating the `ICEBERG` secret, if they're empty we try to look up a secret, if they're not then we don't